### PR TITLE
Add SyncAgentBase

### PR DESCRIPTION
**Pull request by pingsutw**
## Tracking issue
https://github.com/flyteorg/flyte/issues/3936

## Why are the changes needed?
Not all the external services have async API (create/get/delete). For example, Openai API returns the response after the job is completed. Thus, we need to add another API (do) in the agent to run the task synchronously.

## What changes were proposed in this pull request?
- Add the `SyncAgentBase` class for the sync task (ChatGPT task). 
- Rename `AgentBase` to `AsyncAgentBase`
- Add `SyncAgentExecutorMixin` for executing the openAI task locally.

```python
class SyncAgentBase(AgentBase):
    @abstractmethod
    def do(self, task_template: TaskTemplate, inputs: Optional[LiteralMap], **kwargs) -> Resource:
        pass
```

Mock OpenAI agent

```python
class MockAsyncOpenAIAgent(SyncAgentBase):
    name = "mock async openAI Agent"

    def __init__(self):
        super().__init__(task_type_name="async_openai")

    async def do(self, task_template: TaskTemplate, inputs: LiteralMap = None, **kwargs) -> Resource:
        assert inputs.literals["a"].scalar.primitive.integer == 1
        return Resource(phase=TaskExecution.SUCCEEDED, outputs={"o0": 1})
```

I also update the async agent API.

Before:

```python
class AsyncAgentBase(ABC):
    def __init__(self, task_type: str, **kwargs):
        self._task_type = task_type

    def create(
        self,
        output_prefix: str,
        task_template: TaskTemplate,
        inputs: typing.Optional[LiteralMap] = None,
        **kwargs,
    ) -> CreateTaskResponse:  
        ...
        raise CreateTaskResponse(resource_meta=json.dumps(asdict(metadata)).encode("utf-8"))

    def get(self, resource_meta: bytes, **kwargs) -> GetTaskResponse:
        metadata = Metadata(**json.loads(resource_meta.decode("utf-8")))
        raise GetTaskResponse(....)

    def delete(self, resource_meta: bytes, **kwargs) -> DeleteTaskResponse:
        metadata = Metadata(**json.loads(resource_meta.decode("utf-8")))
        raise NotImplementedError
```

After:

```python
class AsyncAgentBase(AgentBase, typing.Generic[T]):
    @abstractmethod
    def create(self, task_template: TaskTemplate, inputs: Optional[LiteralMap], **kwargs) -> ResourceMeta:
        raise NotImplementedError

    @abstractmethod
    def get(self, resource_meta: ResourceMeta, **kwargs) -> Resource:
        raise NotImplementedError

    @abstractmethod
    def delete(self, resource_meta: ResourceMeta, **kwargs):
        raise NotImplementedError
```

## How was this patch tested?
local/remote

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flyte/pull/4874/files

## Docs link
NA

### CODE REVIEW

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -1187,7 +1188,7 @@ def dict_to_literal_map(
         ctx: FlyteContext,
         d: typing.Dict[str, typing.Any],
         type_hints: Optional[typing.Dict[str, type]] = None,
-    ) -> LiteralMap:
+    ) -> Optional[LiteralMap]:
```
Comment (Future-Outlier) [inline flytekit/core/type_engine.py:1191]: Why Optional?

Comment (pingsutw) [reply flytekit/core/type_engine.py]: Revert it. Good catch. I forgot why I made it optional.

File: plugins/flytekit-bigquery/flytekitplugins/bigquery/agent.py [hunk 2]
```diff
@@ -73,54 +63,36 @@ def create(
...
-    def delete(self, resource_meta: bytes, **kwargs) -> DeleteTaskResponse:
+            dst = job.destination
+            if dst:
+                ctx = FlyteContextManager.current_context()
+                output_location = f"bq://{dst.project}:{dst.dataset_id}.{dst.table_id}"
+                res = TypeEngine.dict_to_literal_map(ctx, {"results": StructuredDataset(uri=output_location)})
+
+        return Resource(phase=cur_phase, message=job.state, log_links=[log_link], outputs=res)
+
+    def delete(self, metadata: BigQueryMetadata, **kwargs):
```
Comment (samhita-alla) [inline plugins/flytekit-bigquery/flytekitplugins/bigquery/agent.py:93]: `metadata` needs to be replaced with `resource_meta`

Comment (pingsutw) [reply plugins/flytekit-bigquery/flytekitplugins/bigquery/agent.py]: yes, good catch, thanks. I just updated it.

File: plugins/flytekit-mmcloud/flytekitplugins/mmcloud/agent.py [hunk 3]
```diff
@@ -128,15 +127,12 @@ async def create(
             logger.exception("Cannot open job script for writing")
             raise
 
-        metadata = Metadata(job_id=job_id)
+        return MMCloudMetadata(job_id=job_id)
 
-        return CreateTaskResponse(resource_meta=json.dumps(asdict(metadata)).encode("utf-8"))
-
-    async def async_get(self, resource_meta: bytes, **kwargs) -> GetTaskResponse:
+    async def async_get(self, metadata: MMCloudMetadata, **kwargs) -> Resource:
```
Comment (Future-Outlier) [inline plugins/flytekit-mmcloud/flytekitplugins/mmcloud/agent.py:132]:
```suggestion
    async def get(self, metadata: MMCloudMetadata, **kwargs) -> Resource:
```

File: plugins/flytekit-mmcloud/flytekitplugins/mmcloud/agent.py [hunk 4]
```diff
@@ -173,13 +169,12 @@ async def async_get(self, resource_meta: bytes, **kwargs) -> GetTaskResponse:
         logger.info(f"Obtained status for MMCloud job {job_id}: {job_status}")
         logger.debug(f"OpCenter response: {show_response}")
 
-        return GetTaskResponse(resource=Resource(phase=task_phase))
+        return Resource(phase=task_phase)
 
-    async def async_delete(self, resource_meta: bytes, **kwargs) -> DeleteTaskResponse:
+    async def async_delete(self, metadata: MMCloudMetadata, **kwargs):
```
Comment (Future-Outlier) [inline plugins/flytekit-mmcloud/flytekitplugins/mmcloud/agent.py:174]:
```suggestion
    async def delete(self, metadata: MMCloudMetadata, **kwargs):
```

Comment (Future-Outlier) [reply plugins/flytekit-mmcloud/flytekitplugins/mmcloud/agent.py:174]: mmcloud tests should also be updated.

File: flytekit/extend/backend/base_agent.py [hunk 5]
```diff
@@ -88,186 +178,176 @@ class AgentRegistry(object):
     The agent metadata service will look up the agent metadata based on the agent name.
     """
 
-    _REGISTRY: typing.Dict[str, AgentBase] = {}
-    _METADATA: typing.Dict[str, Agent] = {}
+    _REGISTRY: Dict[TaskCategory, Union[AsyncAgentBase, SyncAgentBase]] = {}
+    METADATA: Dict[str, Agent] = {}
```
Comment (Future-Outlier) [inline flytekit/extend/backend/base_agent.py:182]: not sure why `METADATA` need to keep it public but not private

File: flytekit/extend/backend/base_agent.py [hunk 6]
```diff
@@ -88,186 +178,176 @@ class AgentRegistry(object):
...
+
+        ctx = FlyteContext.current_context()
+        ss = ctx.serialization_settings or SerializationSettings(ImageConfig())
+        task_template = get_serializable(OrderedDict(), ss, self).template
 
-def get_agent_secret(secret_key: str) -> str:
-    return flytekit.current_context().secrets.get(secret_key)
+        agent = AgentRegistry.get_agent(task_template.type, task_template.task_type_version)
+
+        resource = asyncio.run(self._do(agent, task_template, kwargs))
```
Comment (Future-Outlier) [inline flytekit/extend/backend/base_agent.py:245]: Can we improve the error message here?
I have tested it in this PR.
https://github.com/flyteorg/flytekit/pull/2086
Here is the screenshot proof.
image

image


```suggestion
        resource = asyncio.run(self._do(agent, task_template, kwargs))
        if resource.phase != TaskExecution.SUCCEEDED:
            raise FlyteUserException(f"Failed to run the task {agent.name} with error: {resource.message}")
```

**Comment by pingsutw** [comment]
Test Dataflow
image

**Comment by pingsutw** [comment]
Test dataflow

image

**Comment by pingsutw** [comment]
Test BigQuery

image

**Comment by pingsutw** [comment]
bash operator

image

**Comment by pingsutw** [comment]
Databricks

image

**Comment by pingsutw** [comment]
I can't merge it until we cut a flyteidl release.

File: Dockerfile.dev [hunk 7]
```diff
@@ -28,8 +28,8 @@ COPY . /flytekit
 # 3. Clean up the apt cache to reduce image size. Reference: https://gist.github.com/marvell/7c812736565928e602c4
 # 4. Create a non-root user 'flytekit' and set appropriate permissions for directories.
 RUN apt-get update && apt-get install build-essential vim libmagic1 git -y \
+    && pip install "git+https://github.com/flyteorg/flyte.git@master#subdirectory=flyteidl" \
```
Comment (fg91) [inline Dockerfile.dev:31]: <3

**Review by wild-endeavor — APPROVED** [review]

**Pull request merged**